src/Microsoft.SqlTools.ServiceLayer/QueryExecution/DataStorage/SaveAsExcelFileStreamWriterHelper.cs

```diff
@@ -0,0 +1,583 @@
+                writer.WriteStartDocument();
+                writer.WriteStartElement("worksheet", "http://schemas.openxmlformats.org/spreadsheetml/2006/main");
+                writer.WriteAttributeString("xmlns", "r", null, "http://schemas.openxmlformats.org/officeDocument/2006/relationships");
+                writer.WriteStartElement("sheetData");
+                _referenceManager = new ReferenceManager(writer);
+            }
+
+            public void Dispose()
+            {
+                EndRowIfNeeded();
+                _writer.WriteEndElement(); // sheetData 
+                _writer.WriteEndElement(); // worksheet 
+                _writer.Dispose();
+            }
+
+            private XmlWriter _writer;
+            private ReferenceManager _referenceManager;
+        }
+
+        public class ExporterException : Exception
+        {
+            public ExporterException(string message)
+                    : base(message)
+            {
+            }
+        }
+        public SaveAsExcelFileStreamWriterHelper(Stream stream)
+        {
+            zipArchive = new ZipArchive(stream, ZipArchiveMode.Create, true);
+        }
+        public SaveAsExcelFileStreamWriterHelper(Stream stream, bool leaveOpen)
+        {
+            zipArchive = new ZipArchive(stream, ZipArchiveMode.Create, leaveOpen);
+        }
+
+        public ExcelSheet AddSheet(string sheetName = null)
+        {
+            string sheetFileName = "sheet" + (sheetNames.Count + 1);
+            if (sheetName == null)
+            {
+                sheetName = sheetFileName;
+            }
+            EnsureValidSheetName(sheetName);
+
+            sheetNames.Add(sheetName);
+            XmlWriter sheetWriter = AddEntry($"xl/worksheets/{sheetFileName}.xml");
+            return new ExcelSheet(sheetWriter);
+        }
+
+        public void Dispose()
+        {
+            WriteMinimalTemplate();
+            zipArchive.Dispose();
+        }
+
+
+        internal class ReferenceManager
+        {
+            public ReferenceManager(XmlWriter writer)
+            {
+                _writer = writer;
+            }
+            private int _currColumn; // 0 is invalid, the first AddRow will set to 1
+            private int _currRow = 1;
+            private char[] _currReference = new char[3 + 7]; //maximal XFD1048576
+            private int _currReferenceRowLength;
+            private int _currReferenceColumnLength;
+            private XmlWriter _writer;
+
+            public void AssureColumnReference()
+            {
+                if (_currColumn == 0)
+                {
+                    throw new ExporterException("AddRow must be called before AddCell");
+
+                }
+                if (_currColumn > 16384)
+                {
+                    throw new ExporterException("max column number is 16384, see https://support.office.com/en-us/article/Excel-specifications-and-limits-1672b34d-7043-467e-8e27-269d656771c3");
```

Comment:
Can we get away with using InvalidOperationException in these cases rather than define a new exception type?